src/multistream_select/dialer_select.rs

```diff
@@ -346,3 +410,358 @@ impl DialerState {
+                )
+                .await
+                .unwrap()
+                {
+                    Ok((_, io)) => io,
+                    Err(NegotiationError::Failed) => return Ok(()),
+                    Err(NegotiationError::ProtocolError(e)) => {
+                        panic!("Unexpected protocol error {e}")
+                    }
+                };
+                match io.complete().await {
+                    Err(NegotiationError::Failed) => {}
+                    _ => panic!(),
+                }
+
+                Ok(())
+            });
+
+            let client: tokio::task::JoinHandle<Result<(), ()>> = tokio::spawn(async move {
+                let mut io = match tokio::time::timeout(
+                    Duration::from_secs(2),
+                    dialer_select_proto(client_connection, dial_protos, version),
+                )
+                .await
+                .unwrap()
+                {
+                    Err(NegotiationError::Failed) => return Ok(()),
+                    Ok((_, io)) => io,
+                    Err(_) => panic!(),
+                };
+
+                // The dialer may write a payload that is even sent before it
+                // got confirmation of the last proposed protocol, when `V1Lazy`
+                // is used.
+                io.write_all(&dial_payload).await.unwrap();
+                match io.complete().await {
+                    Err(NegotiationError::Failed) => {}
+                    _ => panic!(),
+                }
+
+                Ok(())
+            });
+
+            server.await.unwrap();
+            client.await.unwrap();
+        }
+
+        // Incompatible protocols.
+        run(Version::V1, vec!["/proto1"], vec![1], vec!["/proto2"]).await;
+        run(Version::V1Lazy, vec!["/proto1"], vec![1], vec!["/proto2"]).await;
+    }
+
+    #[tokio::test]
+    async fn v1_lazy_do_not_wait_for_negotiation_on_poll_close() {
+        let (client_connection, _server_connection) =
+            futures_ringbuf::Endpoint::pair(1024 * 1024, 1);
+
+        let client = tokio::spawn(async move {
+            // Single protocol to allow for lazy (or optimistic) protocol negotiation.
+            let protos = vec!["/proto1"];
+            let (proto, mut io) =
+                dialer_select_proto(client_connection, protos, Version::V1Lazy).await.unwrap();
+            assert_eq!(proto, "/proto1");
+
+            // In Libp2p the lazy negotation of protocols can be closed at any time,
+            // even if the negotiation is not yet done.
+
+            // However, for the Litep2p the negotation must conclude before closing the
+            // lazy negotation of protocol. We'll wait for the close until the
+            // server has produced a message, in this test that means forever.
+            io.close().await.unwrap();
+        });
+
+        assert!(tokio::time::timeout(Duration::from_secs(10), client).await.is_err());
```

Comment:
nit: May be add a remark that we should change this to `.is_ok()` and update the comment above after merging https://github.com/paritytech/litep2p/pull/62?